### DISCUSSION TITLE
Refresh shared useAsyncData from outside the page

### DISCUSSION THREAD

**Discussion by Flanker72** [nuxt/nuxt#16575] [Questions]
Hi everyone!

I have a page which fetches some data from the API.
On this page I also have a dynamic component, which needs to trigger refresh of such data.
Tried to make a shared composable with useAsyncData, however there are two different instances of this composable, and `refresh` method does the API request, but main page data still remains the same (it doesn't receive data update).

How can I implement such a behaviour? Passing `refresh` as a component prop is not an option, there may be a number of such handlers.

**Actual result:**

On Button click API request is performed, however {{ data }} is **not** updated. `Updating data from the component` is printed in the console.

**Expected result:**

On Button click API request is performed, {{ data }} is updated. Both `Updating data from the component` and `Updating data from the page` are printed in the console.

**Current project structure:**

`pages/index.ts`:

```
<template>
  <div>
    <div>{{ data }}</div>
    <MyComponent />
  </div>
</template>

<script setup lang="ts">
const { data } = await useMyData();
watch(data, () => {
  console.log("Updating data from the page");
});
</script>
```

`components/MyComponent.vue`:

```
<template>
  <div>
    <button @click="refresh()">Refresh</button>
  </div>
</template>

<script setup lang="ts">
const { data, refresh } = await useMyData();
watch(data, () => {
  console.log("Updating data from the component");
});
</script>
```

`composables/useMyData.ts`:

```
export const useMyData = () => useAsyncData<any[]>("myData", () => {
  return $fetch("/api/data", {
    headers: useRequestHeaders(["cookie"]),
  });
});
```

**Comment by danielroe** [comment]
You can use [`refreshNuxtData`](https://v3.nuxtjs.org/api/utils/refresh-nuxt-data#refreshnuxtdata). Does that look like it will give you what you need?

**Reply by Flanker72** [reply]
Thank for your response!

However, nope. It causes multiple requests in such case.
Problem is that EVERY component, where `useMyData` is imported, triggers fetch after `refreshNuxtData`. It's a real problem, cause I want to have a shared state.

Just to clarify, real business case may looks like this:

1. Site header have a cart indicator, where total items number is displayed.
2. `AddToCart` component triggers cart update and calls `refreshNuxtData("cart")` to refresh cart-related useAsyncData.
3. `AlreadyInCart` component, placed on every product card, should get current cart items to detect if product is already added.

In such a case triggering `refreshNuxtData` produces a number of queries - every `useCart` usage triggers its own fetch, ignoring common cache.

I need to fetch cart data, store it in some kind of a state, and then use this reactive state everywhere within the application.

For now it seems that I need one instance of `useCart` to fetch data, put fetched data into `cartItems` useState, and use different logic to fetch and to read data. It's a mess. I need to fetch data once and use this reactive result everywhere, refetching it only once when it's needed.

Within the one page it works as expected. After splitting over a number of components it leads to inefficient traffic.

`refreshNuxtData` leads to such behavior:
<img width="1679" alt="image" src="https://user-images.githubusercontent.com/619699/174074788-4d54d28e-8ba2-4192-a0af-62d2c68b6556.png">

**Reply by RSS1102** [reply]
> 您可以使用[`refreshNuxtData`](https://v3.nuxtjs.org/api/utils/refresh-nuxt-data#refreshnuxtdata). 看起来它会给你你需要的东西吗？

before my request, I used `refreshNuxtData`, but cross domain problems always occur.
![image](https://user-images.githubusercontent.com/81673017/196072152-247c4702-f819-4b65-b1fa-400cca189fe1.png)
![image](https://user-images.githubusercontent.com/81673017/196072174-ac5e1c59-b66b-4ac7-8eee-9d625ea89e75.png)



If I do not use it, it will not cause cross domain problems, but the requested data will always be the same, even if I carry different parameters.
![image](https://user-images.githubusercontent.com/81673017/196072226-c8e5fe0c-e8ad-4a0e-8dd5-6b3c27d95db1.png)
